### ISSUE TITLE
[BUG] Test runner hangs during regression run in v1.35 

### ISSUE THREAD

**Issue by sarah-gelt** [microsoft/playwright#24071]
### System info
- Playwright Version: [v1.35.1]
- Operating System: [macOS 13.4]
- Browser: [All]
- VSCode: 1.79.2
- node: 16.15.1

### Source code

**Config file**

```js
import { defineConfig, devices } from "@playwright/test";
import dotenv from "dotenv";

/**
 * See https://playwright.dev/docs/test-configuration.
 */

/**
 * Read environment variables from file.
 * https://github.com/motdotla/dotenv
 */
dotenv.config();

export default defineConfig({
  testDir: "./tests",
  projects: [
    { name: "setup", testMatch: ["**/*.setup.ts"], teardown: "teardown" },
    { name: "teardown", testMatch: ["**/*.setup.ts"] },
    {
      name: "chromium",
      use: {
        ...devices["Desktop Chrome"],
        viewport: { width: 1280, height: 768 }
      },
      testMatch: ["**/*.spec.ts"],
      dependencies: ["setup"]
    },
    {
      name: "firefox",
      use: {
        ...devices["Desktop Firefox"],
        viewport: { width: 1280, height: 768 }
      },
      testMatch: ["**/*.spec.ts"],
      dependencies: ["setup"]
    },
    {
      name: "webkit",
      use: {
        ...devices["Desktop Safari"],
        viewport: { width: 1280, height: 768 }
      },
      testMatch: ["**/*.spec.ts"],
      dependencies: ["setup"]
    }
  ],
  /* Shared settings for all projects. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    /* Maximum time each action such as `click()` can take. Defaults to 0 (no limit). */
    actionTimeout: 0,
    /* Base URL to use in actions like `await page.goto('/')`. */
    baseURL: process.env.BASE_URL,
    trace: "retain-on-failure",
    headless: true
    /*launchOptions: {
      slowMo: 100 //Slow test execution down by 100ms, helpful for debugging
    }*/
  },
  /* Maximum time one test can run for. */
  timeout: 2 * 60 * 1000,
  globalTimeout: 25 * 60 * 1000,
  expect: {
    /**
     * Maximum time expect() should wait for the condition to be met.
     * For example in `await expect(locator).toHaveText();`
     */
    timeout: 5000
  },
  maxFailures: process.env.CI ? 16 : undefined,
  /* Run tests in files in parallel */
  fullyParallel: false,
  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,
  retries: process.env.CI ? 3 : 1,
  /* Use 3 works locally. Use workers equal to 50% of available cores on CI - UNTESTED */
  workers: process.env.CI ? "50%" : 3,
  /* Reporters to use. See https://playwright.dev/docs/test-reporters */
  reporter: [
    ["html", { open: "never" }],
    ["./reporters/resultChecker"], //This outputs a file with the current branch and overall run result, to be consumed by GitHub Actions
    ["./reporters/failed-test-reporter.ts"], //This outputs a log of which tests have failed the last time they were executed, so just the failures can be rerun while debugging
    ["./reporters/better-dot-reporter.ts"]    
  ]
});

```

**Steps**
- Upgrade to Playwright 1.35 (previously using 1.34)
- Run our full regression suite via a VSCode terminal using `pnpx playwright test --project chromium` (ours is a monorepo  managed with Rush so we don't use `npx`)

**Expected**
Test run will complete as per usual

**Actual**
The test run stalls part way through, usually close to the end, and doesn't complete. We didn't previously have a `globalTimeout` value set in our config so I added one to help with trying to diagnose this. Once this was set, the runner will abort when that value is exceeded (set to 25 minutes as shown in the config above, a successful run usually takes about 10 minutes) but without the timeout configured I would have to cancel it manually using `ctrl+C` in the VSCode terminal once it become obvious that the runner was no longer responding.

I appreciate this will be a hard one to reproduce as it seems specific to something in our suite. Unfortunately I'm not able to provide our entire repo due to the proprietary information it contains about our application. The problem has occurred for at least two team members (me and one of our developers) so it seems to be common across the board when using 1.35.

I have completed a successful run with 1.34 and a hung/aborted run with 1.35 both using `DEBUG=pw:api` and saved output the logs to file. The files include bearer tokens so I would rather not attach them here for security reasons but would be very happy to email them direct if they will potentially help with identifying a cause.

Dot reporter output from most recent run which was aborted by the global timeout:
```
Running 158 tests using 3 workers
··········································································°·····
···················°······°°······································°··°×T
Timed out waiting 1500s for the test suite to run

Timed out waiting 1500s for the teardown for start workers to run


  1) [chromium] › 04_managedentities.spec.ts:894:9 › Managed Entities › Create & manage new Legal Entity › ME15 - Deleting a business entity with ownership stake in another business entity @notvisual 

    "afterAll" hook timeout of 120000ms exceeded.

      969 |     });
      970 |
    > 971 |     test.afterAll(async () => {
          |          ^
      972 |       await deleteAllLegalEntities();
      973 |     });
      974 |   });

        at /Users/sarahwoodhouse/Documents/repos/core-platform/apps/gelt-frontend-autotests/tests/04_managedentities.spec.ts:971:10

    Retry #1 ───────────────────────────────────────────────────────────────────────────────────────

    "afterAll" hook timeout of 120000ms exceeded.

      969 |     });
      970 |
    > 971 |     test.afterAll(async () => {
          |          ^
      972 |       await deleteAllLegalEntities();
      973 |     });
      974 |   });

        at /Users/sarahwoodhouse/Documents/repos/core-platform/apps/gelt-frontend-autotests/tests/04_managedentities.spec.ts:971:10

  Slow test file: [chromium] › 01_taxprofile.spec.ts (4.7m)
  Slow test file: [chromium] › 03_docvault.spec.ts (3.7m)
  Slow test file: [chromium] › 05_assets.cryptoaccount.spec.ts (2.2m)
  Slow test file: [chromium] › 02_assets.privateinvestment.spec.ts (2.0m)
  Slow test file: [chromium] › 04_managedentities.spec.ts (1.4m)
  Consider splitting slow test files to speed up parallel execution
  1 failed
    [chromium] › 04_managedentities.spec.ts:894:9 › Managed Entities › Create & manage new Legal Entity › ME15 - Deleting a business entity with ownership stake in another business entity @notvisual 
  13 skipped
  144 passed (25.0m)
  2 errors were not a part of any test, see above for details
  ```

**Comment by pavelfeldman** [comment]
It is likely that the newer browser that comes with 1.35 fails your test. Could you narrow down which test and which call is affected? Try recording the trace to see what's happening in the browser.

**Comment by sarah-gelt** [comment]
I've been able to identify four tests in one class that seem to trigger this problem. When it happens, the node process that is being used by the test runner hits around 100% CPU usage according to the macOS Activity Monitor and gets stuck there. I have to force stop that process after the test run is aborted by the global timeout, which I've reduced to 2 mins now I'm able to reproduce this with one test.

The problem actually seems to occur at the end of the `afterAll` hook. I've put console logs in the hook including as the very last line within the block and all the logs are output to the terminal. It seems to be internal to the runner, when it tries to close the hook before moving on to the global teardown. The last line of my code (a `console.log()`) in the `afterAll` is run but then the runner it never makes it to the first line of my code in the teardown.

Runner output:
```DEBUG=pw:api pnpx playwright test --project chromium --reporter=dot 2>debug.log

Running 3 tests using 1 worker
·
Timed out waiting 120s for the test suite to run

Timed out waiting 120s for the teardown for start workers to run

  2 skipped
  1 passed (2.0m)
  Timed out waiting 120s for the entire test run
  2 errors were not a part of any test, see above for details
```

The report supports this, suggesting that the afterAll is where the test failed but in fact showing every step of it as passing.
<img width="1016" alt="image" src="https://github.com/microsoft/playwright/assets/107568761/f21184df-4762-4c3f-923e-03bdfc00df6c">
Report attached as it's too long to easily screenshot all the `afterAll` steps
[Playwright Test Report.html.zip](https://github.com/microsoft/playwright/files/11981375/Playwright.Test.Report.html.zip)

What's especially strange is that most of the tests in this class pass, so the `afterAll` hook can be completed correctly for 10 of the 14 tests in the class.

**Comment by sarah-gelt** [comment]
Again I have recorded the DEBUG=pw:api output for this specific test with 1.35 where the run gets stalled and with 1.34 where the run succeeds. I'm happy to provide them directly via email if it would be useful.

**Comment by alexanderdevm** [comment]
I am having a similar issue with 1.35.1 version of playwright, where IT would hang indefinitely especially if chromium is used and executed against several projects, even though tests are mainly api and/or over ssh2 lib.

What i have noticed is at one point after ssh2.exec command would finish, it would hang.
This behaviour is really noticeable when tests run in playwright docker.

I was not able to get playwright to give any traces or error messages other than (this point of hanging):
```
message: '\x1BWorker teardown timeout of 30000ms exceeded while tearing down "trace recording".\x1B\n' +

[17:58:30 ](...
      '\x1BFailed worker ran 4 tests:\x1B\n' +
```

( issue is not present in playwright 1.34.3)

playwright.config.ts
```
import { devices, PlaywrightTestConfig } from '@playwright/test';


const config: PlaywrightTestConfig = {
  testDir: './tests',
  /* Maximum time one test can run for. */
  timeout: 30 * 1000,
  expect: {
    timeout: 50000
  },
  /* Run tests in files in parallel */
  fullyParallel: true,
  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,
  /* Retry on CI only */
  retries: process.env.CI ? 2 : 0,
  /* Opt out of parallel tests on CI. */
  workers: process.env.CI ? 1 : undefined,
  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  reporter: process.env.CI
    ? [
        [
          'playwright-teamcity-reporter',
          { testMetadataArtifacts: 'test-results', logConfig: false }
        ],
        ['html', { open: 'never' }]
      ]
    : [
        ['html', { open: 'never' }],
        ['list', { printSteps: true }]
      ],
  /* Shared settings for all the projects below.
    See https://playwright.dev/docs/api/class-testoptions. */
  globalSetup: require.resolve('./tests/global-setup'),
  globalTeardown: require.resolve('./tests/global-teardown'),
  use: {
    storageState: COOKIE_PATH,
    headless: true,
    actionTimeout: 0,
    trace: 'retain-on-failure',
    ignoreHTTPSErrors: true,
    baseURL: ...,
    extraHTTPHeaders: { Cookie: `id=${...}` },
    launchOptions: {
      slowMo: parseInt(process.env.PLAYWRIGHT_SLOW_MO || '0')
    }
  },

  projects: [
    {
      name: 'chromium',
      use: {
        ...devices['Desktop Chrome']
      }
    },
    {
      name: 'api',
      testDir: './tests/api',
      testMatch: /.\/tests\/api\/.*.spec.ts/,
      use: {
        headless: true,
        contextOptions: {
          ignoreHTTPSErrors: true
        }
      }
    },
    {
      name: 'ui',
      testDir: './tests/ui',
      testMatch: /.\/tests\/ui\/.*.spec.ts/,
      use: {
        ...devices['Desktop Chrome'],
      }
    },

  webServer:
    project !== 'api'
      ? [
          {
            command: 'yarn start',
            url: '...',
            ignoreHTTPSErrors: true,
            reuseExistingServer: true
          }
        ]
      : []
};


}

export default config;
```

**Comment by alexanderdevm** [comment]
I have enabled debugging on both ssh lib and playwright api

```
Inbound: Received USERAUTH_FAILURE (publickey,password)
Client: none auth failed
Outbound: Sending USERAUTH_REQUEST (password)
Inbound: Received USERAUTH_SUCCESS
Outbound: Sending CHANNEL_OPEN (r:0, session)
Inbound: GLOBAL_REQUEST (hostkeys-00@openssh.com)
Outbound: Sending GLOBAL_REQUEST (hostkeys-prove-00@openssh.com)
Inbound: CHANNEL_OPEN_CONFIRMATION (r:0, s:0)
Outbound: Sending CHANNEL_REQUEST (r:0, exec: sudo -S -p "...")
Inbound: REQUEST_SUCCESS
Inbound: CHANNEL_WINDOW_ADJUST (r:0, 2097152)
Inbound: CHANNEL_SUCCESS (r:0)
Inbound: CHANNEL_EXTENDED_DATA (r:0, 28)
Outbound: Sending CHANNEL_DATA (r:0, 8)
Inbound: CHANNEL_EOF (r:0)
Inbound: CHANNEL_REQUEST (r:0, exit-status: 0)
Inbound: CHANNEL_CLOSE (r:0)
Outbound: Sending CHANNEL_CLOSE (r:0)
Outbound: Sending DISCONNECT (11)
Outbound: Sending DISCONNECT (11)
Socket ended
Socket closed
Socket ended
Socket closed
  pw:api => apiRequestContext.storageState started +17s
  pw:api <= apiRequestContext.storageState succeeded +2ms
  pw:api => apiRequestContext.get started +1ms
  pw:api → GET https://.../v2/inputs +4ms
  pw:api   user-agent: Playwright/1.35.1 (x64; ubuntu 22.04) node/16.19 +0ms
  pw:api   accept: application/json +0ms
  pw:api   accept-encoding: gzip,deflate,br +0ms
  pw:api   Cookie: id=... +0ms
  pw:api ← 200 OK +10ms
  pw:api   server: nginx +0ms
  pw:api   date: Sat, 08 Jul 2023 05:25:05 GMT +0ms
  pw:api   content-type: application/json +0ms
  pw:api   transfer-encoding: chunked +0ms
  pw:api   connection: close +0ms
  pw:api   vary: Accept-Encoding +0ms
  pw:api   cache-control: no-cache, no-store, must-revalidate +0ms
  pw:api   pragma: no-cache +0ms
  pw:api   expires: 0 +0ms
  pw:api   strict-transport-security: max-age=31536000; includeSubDomains +0ms
  pw:api   x-content-type-options: nosniff +0ms
  pw:api   x-frame-options: SAMEORIGIN +0ms
  pw:api   x-xss-protection: 1; mode=block +0ms
  pw:api   content-encoding: gzip +0ms
  pw:api <= apiRequestContext.get succeeded +3ms
  pw:api => apiResponse.text started +2ms
  pw:api <= apiResponse.text succeeded +1ms
  pw:api => apiResponse.json started +3ms
  pw:api <= apiResponse.json succeeded +1ms
  pw:api => tracing.stopChunk started +19ms
  pw:api => tracing.stopChunk started +1ms
  pw:api <= tracing.stopChunk succeeded +0ms
Worker teardown timeout of 30000ms exceeded while tearing down "trace recording".
```

**Comment by agataciesielska** [comment]
I have exactly the same situation. The run is stucked even if afterAll hook is empty (sic!)
![image](https://github.com/microsoft/playwright/assets/11655108/504c138c-ffbf-4ec7-ba47-f336d9a825fa)
Downgrading to 1.32.2 playwright version seems to be free of this issue.

**Comment by mxschmitt** [comment]
Hey folks! We unfortunately need a clear reproducible in order to act on this. Since this issue most likely involves a bunch of different internal parts, logs are not enough for us to solve the puzzle.

Would it be possible to share a test-suite? You can also share it privately with me if you have narrowed it down to e.g. a single test file already. `max` `schmitt ` `@` `microsoft.com`. Thanks!

**Comment by alexanderdevm** [comment]
> Hey folks! We unfortunately need a clear reproducible in order to act on this. Since this issue most likely involves a bunch of different internal parts, logs are not enough for us to solve the puzzle.
> 
> Would it be possible to share a test-suite? You can also share it privately with me if you have narrowed it down to e.g. a single test file already. `max` `schmitt ` `@` `microsoft.com`. Thanks!

@mxschmitt while there is a possibility to share the code, do you need the test suite to be in running state ?  the environment its running against is local only.

**Comment by mxschmitt** [comment]
Yes, the test suite needs to be in a running state. Ideally we run `npx playwright test` and it hangs.

**Comment by mxschmitt** [comment]
Yes it needs to be in a running state. You can also try to reduce the test case to a smaller one which replicates your issue, so its less you need to share / setup.

**Comment by aslushnikov** [comment]
Unfortunately this is not actionable unless we have a good and clear reproducible. I'll close this for now, but please do not hesitate to open a new issue if you have a good repro.

Thanks!

**Comment by peterpeterparker** [comment]
@aslushnikov, I don't have a sample repository to share, but our design system built with SvelteKit is relatively simple, and we are facing this problem.

I've just opened a [PR](https://github.com/dfinity/gix-components/pull/257) to reproduce the problem.

You can see the CI results there: https://github.com/dfinity/gix-components/actions/runs/5645052302/job/15290064832?pr=257

As you can see, the CI hangs forever. Let me know if this information helps in identifying the issue.

**Comment by aslushnikov** [comment]
@peterpeterparker I can repro; this is fixed on tip-of-tree with https://github.com/microsoft/playwright/pull/24242

The fix will go live with Playwright 1.37 (tentatively scheduled for sometime next week). Meanwhile, you can try with @playwright/test@next and see if it works for you (works for me!).

**Comment by peterpeterparker** [comment]
Thanks @aslushnikov! Can confirm the fix. I bumped next version in my branch and CI went through ([GitHub action](https://github.com/dfinity/gix-components/actions/runs/5681128998/job/15396616806?pr=257) for refence).

**Comment by aslushnikov** [comment]
> Unfortunately not good news from my side. 

@sarah-gelt Sad to hear.

> Is there any further debugging etc I can do to try to track down the issue?

The best way would be to gradually remove code from your tests and fixtures, until you figure our exactly what makes the test runner hang.

**Comment by sarah-gelt** [comment]
@aslushnikov After a long afternoon of picking things apart, I think I've found the cause. It seems to relate to how many API calls the suite is making. Whenever I finish using an instance of `APIRequestContext` I now call `.dispose()` and this has resolved the issue for me. Perhaps there were too many responses in memory that couldn't get cleared down at the end of the run? This particular class does a lot of creating and deleting records in the system via the API to support the front end validation, so seems particularly affected.
